pact/provider/core/src/main/java/org/arquillian/algeron/pact/provider/core/InteractionRunner.java

```diff
@@ -224,18 +208,104 @@ protected void executeStateChanges(final RequestResponseInteraction interaction,
         if (!interaction.getProviderStates().isEmpty()) {
             for (final ProviderState state : interaction.getProviderStates()) {
                 Arrays.stream(testClass.getMethods(State.class))
-                        .filter(method -> ArrayUtils.contains(method.getAnnotation(State.class).value(), state.getName()))
+                        .filter(method -> ArrayUtils.matches(
+                                method.getAnnotation(State.class).value(), state.getName()))
                         .forEach(method -> {
-                            if (method.getParameterCount() == 1) {
+                            if (isStateMethodWithMapParameter(method)) {
                                 executeMethod(method, target, state.getParams());
                             } else {
-                                executeMethod(method, target);
+                                if (method.getParameterCount() > 0) {
+                                    // Use regular expressions to pass parameters.
+                                    executeStateMethodWithRegExp(method, state, target);
+                                } else {
+                                    executeMethod(method, target);
+                                }
                             }
                         });
             }
         }
     }
 
+    private void executeStateMethodWithRegExp(Method stateMethod, ProviderState state, Object target) {
+        final Optional<String> matchingStateOptional = ArrayUtils.firstMatch(
+                stateMethod.getAnnotation(State.class).value(), state.getName());
+        // We are sure that at this point, an state is present
+        final String matchingState = matchingStateOptional.get();
+        final List<String> arguments = ArgumentPatternMatcher.arguments(
+                Pattern.compile(matchingState), state.getName());
+
+        if (arguments.size() != stateMethod.getParameterCount()) {
+            throw new IllegalArgumentException(String.format("Consumer state %s matches with provider state %s but provider method contains %s arguments instead of matching %s",
+                    state.getName(), matchingState, stateMethod.getParameterCount(), arguments.size()));
+        }
+
+        Class<?>[] parameterTypes = stateMethod.getParameterTypes();
+
+        final Object[] instances = new Object[parameterTypes.length];
+
+        for (int i = 0; i < parameterTypes.length; i++) {
+            Class<?> parameter = parameterTypes[i];
+            if (isSimpleType(parameter)) {
+                instances[i] = cast(arguments.get(i), parameter);
+            } else {
+                if (isCollectionType(parameter)) {
+                    instances[i] = cast(arguments.get(i));
+                }
+            }
+        }
+
+        executeMethod(stateMethod, target, instances);
+    }
+
+
+    private Collection<String> cast(String argument) {
```

Comment:
I'm not sure if type conversion is a responsibility of `InteractionRunner.java`. How about moving it to dedicated component and test in isolation there? You might also want to have a look at what's implement in APE for this kind of typical string-type conversion https://github.com/arquillian/arquillian-extension-persistence/blob/master/core/src/main/java/org/jboss/arquillian/persistence/core/configuration/ConfigurationTypeConverter.java